File: src/banner/text_style.py
"""Text style extraction from image regions."""

from __future__ import annotations

import base64
from collections import Counter
from io import BytesIO
from json import loads
from typing import TYPE_CHECKING

import numpy as np
from PIL import Image
from sklearn.cluster import KMeans

from .models import TextStyle

if TYPE_CHECKING:
    from openai import OpenAI

RGB = tuple[int, int, int]

STYLE_PROMPT = """
Describe the text styling in this image. Respond in the form of a JSON object with keys for the text
foreground color, the text border/drop shadow color, and approximate font weight. Provide the colors
as hex codes, and font weight as a number from 100-900. Respond only with JSON. Do not analyze, do
it yourself.
"""

STYLE_RESPONSE_FORMAT = {
    "type": "json_schema",
    "json_schema": {
        "name": "text_style",
        "schema": {
            "type": "object",
            "properties": {
                "foreground_color": {
                    "type": "string",
                    "description": "The foreground color of the text",
                },
                "border_color": {
                    "type": "string",
                    "description": "The text border or drop shadow color.",
                },
                "font_weight": {
                    "type": "integer",
                    "description": "The approximate font weight of the text",
                    "minimum": 100,
                    "maximum": 900,
                },
            },
            "required": ["foreground_color", "border_color", "font_weight"],
        },
    },
}


def image_to_data_uri(image: Image.Image) -> str:
    """Convert PIL Image to data URI."""
    buffer = BytesIO()
    image.save(buffer, format="PNG")
    img_str = base64.b64encode(buffer.getvalue())
    return f"data:image/png;base64,{img_str.decode('utf-8')}"


def get_dominant_colors(image: Image.Image, num_colors: int = 5) -> list[RGB]:
    """Extract dominant colors from an image using K-means clustering."""
    img = image.convert("RGB")
    img_array = np.array(img)
    pixels = img_array.reshape(-1, 3)

    kmeans = KMeans(n_clusters=num_colors, n_init=10)
    kmeans.fit(pixels)

    colors = kmeans.cluster_centers_.astype(int)
    labels = kmeans.labels_
    counts = Counter(labels)
    total_pixels = sum(counts.values())

    color_percentages = [
        (tuple(int(c) for c in color), count / total_pixels * 100)
        for color, count in zip(colors, [counts[i] for i in range(num_colors)])
    ]

    color_percentages.sort(key=lambda x: x[1], reverse=True)
    return [color[0] for color in color_percentages]


def closest_color(colors: list[RGB], target: RGB) -> RGB:
    """Find the closest color in a list to the target color."""
    colors_arr = np.array(colors)
    target_arr = np.array(target)
    distances = np.sqrt(np.sum((colors_arr - target_arr) ** 2, axis=1))
    index = np.argmin(distances)
    return tuple(int(c) for c in colors_arr[index])


def rgb_to_hex(rgb: RGB) -> str:
    """Convert RGB tuple to hex color string."""
    return "#{:02x}{:02x}{:02x}".format(rgb[0], rgb[1], rgb[2])


def hex_to_rgb(hexcode: str) -> RGB:
    """Convert hex color string to RGB tuple."""
    return tuple(bytes.fromhex(hexcode.lstrip("#")))


class TextStyleExtractor:
    """Extracts text styling from image regions using OpenAI vision."""

    def __init__(self, client: OpenAI, model: str = "gpt-4o"):
        self.client = client
        self.model = model

    def _get_raw_style(self, image: Image.Image) -> dict:
        """Get raw text style from OpenAI."""
        image_data_uri = image_to_data_uri(image)

        completion = self.client.chat.completions.create(
            model=self.model,
            response_format=STYLE_RESPONSE_FORMAT,
            messages=[
                {"role": "system", "content": STYLE_PROMPT},
                {
                    "role": "user",
                    "content": [
                        {"type": "image_url", "image_url": {"url": image_data_uri}}
                    ],
                },
            ],
        )
        return loads(completion.choices[0].message.content)

    def extract(self, image: Image.Image) -> TextStyle:
        """
        Extract text style from an image region.

        Uses OpenAI to identify colors, then snaps them to dominant
        colors in the image for accuracy.
        """
        colors = get_dominant_colors(image, num_colors=5)
        raw_style = self._get_raw_style(image)

        # Snap colors to dominant colors in the image
        fg_rgb = hex_to_rgb(raw_style["foreground_color"])
        foreground_color = rgb_to_hex(closest_color(colors, fg_rgb))

        border_rgb = hex_to_rgb(raw_style["border_color"])
        border_color = rgb_to_hex(closest_color(colors, border_rgb))

        return TextStyle(
            foreground_color=foreground_color,
            border_color=border_color,
            font_weight=raw_style["font_weight"],
            font_size=image.height * 0.8,
        )

File: workers/banner/src/text-style.ts
/**
 * Text style extraction from image regions.
 */

import OpenAI from "openai";
import type { TextStyle, RGB, BoundingBox } from "./types";
import { getRectifiedDimensions } from "./image-utils";

const STYLE_PROMPT = `
Describe the text styling in this image. Respond in the form of a JSON object with keys for the text
foreground color, the text border/drop shadow color, and approximate font weight. Provide the colors
as hex codes, and font weight as a number from 100-900. Respond only with JSON. Do not analyze, do
it yourself.
`;

const STYLE_RESPONSE_FORMAT = {
	type: "json_schema" as const,
	json_schema: {
		name: "text_style",
		schema: {
			type: "object",
			properties: {
				foreground_color: {
					type: "string",
					description: "The foreground color of the text",
				},
				border_color: {
					type: "string",
					description: "The text border or drop shadow color.",
				},
				font_weight: {
					type: "integer",
					description: "The approximate font weight of the text",
					minimum: 100,
					maximum: 900,
				},
			},
			required: ["foreground_color", "border_color", "font_weight"],
		},
	},
};

/**
 * Simple K-means clustering implementation.
 */
function kMeans(
	points: number[][],
	k: number,
	maxIterations: number = 10
): { centroids: number[][]; labels: number[] } {
	if (points.length === 0) {
		return { centroids: [], labels: [] };
	}

	const dim = points[0].length;

	// Initialize centroids randomly from points
	const indices = new Set<number>();
	while (indices.size < k && indices.size < points.length) {
		indices.add(Math.floor(Math.random() * points.length));
	}
	let centroids = Array.from(indices).map((i) => [...points[i]]);

	let labels = new Array(points.length).fill(0);

	for (let iter = 0; iter < maxIterations; iter++) {
		// Assign points to nearest centroid
		const newLabels = points.map((point) => {
			let minDist = Infinity;
			let minLabel = 0;
			for (let c = 0; c < centroids.length; c++) {
				let dist = 0;
				for (let d = 0; d < dim; d++) {
					dist += Math.pow(point[d] - centroids[c][d], 2);
				}
				if (dist < minDist) {
					minDist = dist;
					minLabel = c;
				}
			}
			return minLabel;
		});

		// Check for convergence
		let changed = false;
		for (let i = 0; i < labels.length; i++) {
			if (labels[i] !== newLabels[i]) {
				changed = true;
				break;
			}
		}
		labels = newLabels;

		if (!changed) break;

		// Update centroids
		const newCentroids: number[][] = [];
		for (let c = 0; c < k; c++) {
			const clusterPoints = points.filter((_, i) => labels[i] === c);
			if (clusterPoints.length > 0) {
				const centroid = new Array(dim).fill(0);
				for (const point of clusterPoints) {
					for (let d = 0; d < dim; d++) {
						centroid[d] += point[d];
					}
				}
				for (let d = 0; d < dim; d++) {
					centroid[d] /= clusterPoints.length;
				}
				newCentroids.push(centroid);
			} else {
				newCentroids.push(centroids[c]);
			}
		}
		centroids = newCentroids;
	}

	return { centroids, labels };
}

/**
 * Extract dominant colors from image pixel data using K-means clustering.
 *
 * Note: In Cloudflare Workers we can't decode the actual pixels from a base64 image
 * without additional libraries. This function is provided for completeness but
 * the style extractor will rely primarily on the OpenAI vision model.
 */
export function getDominantColors(pixels: RGB[], numColors: number = 5): RGB[] {
	if (pixels.length === 0) {
		return [];
	}

	const points = pixels.map((rgb) => [rgb[0], rgb[1], rgb[2]]);
	const { centroids, labels } = kMeans(points, numColors);

	// Count pixels per cluster
	const counts = new Map<number, number>();
	for (const label of labels) {
		counts.set(label, (counts.get(label) || 0) + 1);
	}

	// Sort by frequency
	const indexed = centroids.map((c, i) => ({
		color: c,
		count: counts.get(i) || 0,
	}));
	indexed.sort((a, b) => b.count - a.count);

	return indexed.map(
		(c) =>
			[
				Math.round(c.color[0]),
				Math.round(c.color[1]),
				Math.round(c.color[2]),
			] as RGB
	);
}

/**
 * Find the closest color in a list to the target color.
 */
export function closestColor(colors: RGB[], target: RGB): RGB {
	let minDist = Infinity;
	let closest: RGB = colors[0] || [0, 0, 0];

	for (const color of colors) {
		const dist = Math.sqrt(
			Math.pow(color[0] - target[0], 2) +
				Math.pow(color[1] - target[1], 2) +
				Math.pow(color[2] - target[2], 2)
		);
		if (dist < minDist) {
			minDist = dist;
			closest = color;
		}
	}

	return closest;
}

/**
 * Convert RGB tuple to hex color string.
 */
export function rgbToHex(rgb: RGB): string {
	return (
		"#" +
		rgb
			.map((c) => Math.max(0, Math.min(255, c)).toString(16).padStart(2, "0"))
			.join("")
	);
}

/**
 * Convert hex color string to RGB tuple.
 */
export function hexToRgb(hex: string): RGB {
	const h = hex.replace("#", "");
	return [
		parseInt(h.substring(0, 2), 16),
		parseInt(h.substring(2, 4), 16),
		parseInt(h.substring(4, 6), 16),
	];
}

interface RawStyleResponse {
	foreground_color: string;
	border_color: string;
	font_weight: number;
}

/**
 * Extracts text styling from image regions using OpenAI vision.
 */
export class TextStyleExtractor {
	private client: OpenAI;
	private model: string;

	constructor(apiKey: string, model: string = "gpt-4o") {
		this.client = new OpenAI({ apiKey });
		this.model = model;
	}

	/**
	 * Get raw text style from OpenAI.
	 */
	private async getRawStyle(imageDataUri: string): Promise<RawStyleResponse> {
		const completion = await this.client.chat.completions.create({
			model: this.model,
			response_format: STYLE_RESPONSE_FORMAT,
			messages: [
				{ role: "system", content: STYLE_PROMPT },
				{
					role: "user",
					content: [
						{ type: "image_url", image_url: { url: imageDataUri } },
					],
				},
			],
		});

		const content = completion.choices[0]?.message?.content;
		if (!content) {
			throw new Error("No style response from OpenAI");
		}

		return JSON.parse(content) as RawStyleResponse;
	}

	/**
	 * Extract text style from an image region.
	 *
	 * Uses OpenAI to identify colors and font weight.
	 * Since we can't extract pixels in Cloudflare Workers, we trust the
	 * OpenAI vision model's color identification directly.
	 *
	 * @param imageDataUri - Base64 data URI of the text region image
	 * @param box - The bounding box for calculating font size
	 * @returns TextStyle with colors and font info
	 */
	async extract(imageDataUri: string, box: BoundingBox): Promise<TextStyle> {
		const rawStyle = await this.getRawStyle(imageDataUri);
		const { height } = getRectifiedDimensions(box);

		return {
			foregroundColor: rawStyle.foreground_color,
			borderColor: rawStyle.border_color,
			fontWeight: rawStyle.font_weight,
			fontSize: height * 0.8,
		};
	}
}
